### PR TITLE
S2U-34 Assignments add rubric support for peer review

### PR DESCRIPTION
**Pull request by stetsche**
https://sakaiproject.atlassian.net/browse/S2U-34

Conversion script: https://github.com/sakaiproject/sakai-reference/pull/222

### CODE REVIEW

**Comment by bgarciaentornos** [comment]
Thanks! 5 years since the original development and after more than a few updates, this might make it to master!

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentPeerAssessmentServiceImpl.java [hunk 1]
```diff
@@ -56,10 +63,11 @@
...
 
-    private ScheduledInvocationManager scheduledInvocationManager;
-    protected AssignmentService assignmentService;
-    private SecurityService securityService = null;
-    private SessionManager sessionManager;
+    @Getter @Setter private ScheduledInvocationManager scheduledInvocationManager;
+    @Getter @Setter protected AssignmentService assignmentService;
+    @Getter @Setter private SecurityService securityService = null;
+    @Getter @Setter private SessionManager sessionManager;
+    @Getter @Setter private SiteService siteService;
```
Comment (ern) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentPeerAssessmentServiceImpl.java:70]: most of the time you only need Setter for services

Comment (stetsche) [reply assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentPeerAssessmentServiceImpl.java:70]: Yes, I think we can remove the getters, right @bgarciaentornos?

Comment (mpellicer) [reply assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentPeerAssessmentServiceImpl.java:70]: Hey Markus @stetsche, I think we can remove them, however, I'm not a big fan of changing code that is working and has been verified, even if they are simple and small changes. Let's take advantage of the effort spent on this, this is running in production for many years right now.

I prefer to not introduce differences between Master and the 22.x version of the feature, we did that in the past and that practices almost killed all of our rubrics features, this has been around for five years and we want to be consistent between versions.

Let's stick with what it is in 22.x please.

**Comment by mpellicer** [comment]
Does this need a conversion script for 24?

**Comment by stetsche** [comment]
It seems to work without, but I'll create a PR for it as well.

**Comment by bgarciaentornos** [comment]
It does need a conversion script, forgot to mention it initially

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 2]
```diff
@@ -3819,8 +3834,24 @@ protected String build_instructor_grade_submission_context(VelocityPortlet portl
...
 
             boolean allowToGrade = true;
-            if (StringUtils.isNotBlank(a.getProperties().get(PROP_ASSIGNMENT_ASSOCIATE_GRADEBOOK_ASSIGNMENT))) {
+            String assignmentAssociateGradebook = a.getProperties().get(PROP_ASSIGNMENT_ASSOCIATE_GRADEBOOK_ASSIGNMENT);
+            if (StringUtils.isNotBlank(assignmentAssociateGradebook)) {
                 String gradebookUid = toolManager.getCurrentPlacement().getContext();
+                // If the assignment reference is not equal to the associated gradebook item, then a custom gb item is being used
+                if (!assignmentRef.equals(assignmentAssociateGradebook)) {
+                    try {
+                        org.sakaiproject.grading.api.Assignment gbAssignment = gradingService.getAssignment(gradebookUid,
```
Comment (mpellicer) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:3843]: Please put a comment about what differences are between Master and 22.x, what lines did we replace. Thanks

**Comment by mpellicer** [comment]
Please put comments in the relevant parts where there are differences between 22.x and Master.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 2 above)
Comment (stetsche) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:3843]: The logic between master and 22 here should work the same.
On 22.x an `org.sakaiproject.service.gradebook.shared.Assignment` object is retrieved from the `GradebookService` if it's null we don't set the context variables.
For the master an `org.sakaiproject.service.gradebook.shared.Assignment` object is retrieved from the `GradingService`, if an `AssessmentNotFoundException` (yes Assessment) thrown, the variables are not added to the context.

**Review by stetsche — COMMENTED** [review]
@mpellicer I added a few comments related to the migration.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 2 above)
Comment (mpellicer) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:3843]: But put a comment int the source! :D

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 3]
```diff
@@ -3819,8 +3834,24 @@ protected String build_instructor_grade_submission_context(VelocityPortlet portl
             state.setAttribute(NEW_ASSIGNMENT_CHECK_ANONYMOUS_GRADING, assignmentService.assignmentUsesAnonymousGrading(a));
 
             boolean allowToGrade = true;
-            if (StringUtils.isNotBlank(a.getProperties().get(PROP_ASSIGNMENT_ASSOCIATE_GRADEBOOK_ASSIGNMENT))) {
+            String assignmentAssociateGradebook = a.getProperties().get(PROP_ASSIGNMENT_ASSOCIATE_GRADEBOOK_ASSIGNMENT);
+            if (StringUtils.isNotBlank(assignmentAssociateGradebook)) {
                 String gradebookUid = toolManager.getCurrentPlacement().getContext();
+                // If the assignment reference is not equal to the associated gradebook item, then a custom gb item is being used
```
Comment (stetsche) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:3840]:
```suggestion
                // If the assignment reference is not equal to the associated gradebook item, then a custom gb item is being used
                // S2U-34 In the 22x version this check was looking if the retrieved object from the gradingservice is null, now we check if an exception is thrown
```

Comment (stetsche) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:3840]: @mpellicer

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubrics-api-mixin.js [hunk 4]
```diff
@@ -30,7 +30,12 @@ export const rubricsApiMixin = Base => class extends Base {
 
   apiGetEvaluation() {
 
-    const url = `/api/sites/${this.siteId}/rubric-evaluations/tools/${this.toolId}/items/${this.entityId}/evaluations/${this.evaluatedItemId}`;
```
[the remark this answers is not in the document]
Comment (stetsche) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubrics-api-mixin.js:33]: I don't think this needs a comment in the source.

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java [hunk 5]
```diff
@@ -862,14 +870,10 @@ public EvaluationTransferBean cancelDraftEvaluation(Long draftEvaluationId) {
...
     public boolean hasAssociatedRubric(String tool, String id) {
-        return hasAssociatedRubric(tool, id, toolManager.getCurrentPlacement().getContext());
-    }
-
-    public boolean hasAssociatedRubric(String tool, String id, String siteId) {
 
         if (StringUtils.isBlank(id)) return false;
 
-        return associationRepository.findByToolIdAndItemId(tool, id).filter(canEvaluate.or(canEdit).or(isCreator)).isPresent();
+        return getRubricAssociation(tool, id).isPresent();
```
Comment (stetsche) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java:876]: @mpellicer 
```suggestion
        // S2U-34 This is bypassing the permission check as it's needed for peer review
        return getRubricAssociation(tool, id).isPresent();
```

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 6]
```diff
@@ -3819,8 +3834,25 @@ protected String build_instructor_grade_submission_context(VelocityPortlet portl
...
+                // S2U-34 In the 22x version this check was looking if the retrieved object from the gradingservice is null, now we check if an exception is thrown
+                if (!assignmentRef.equals(assignmentAssociateGradebook)) {
+                    try {
+                        org.sakaiproject.grading.api.Assignment gbAssignment = gradingService.getAssignment(gradebookUid,
+                                assignmentAssociateGradebook);
+
+                        Long associateGradebookAssignmentId = gbAssignment.getId();
+                        context.put("associatedToGbItem", true);
+                        context.put("associatedToGbEntityId", associateGradebookAssignmentId);
+                    } catch (AssessmentNotFoundException e) {
```
Comment (mpellicer) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:3850]: wtf

**Review by mpellicer — APPROVED** [review]

**Comment by mpellicer** [comment]
Terrific work guys, I want to thank you all involved in this piece of work for five years, it's been a long journey but it's definitely contributed. @bgarciaentornos @stetsche @MRutea @josecebe and Manu Fuster.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 6 above)
Comment (adrianfish) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:3850]: AssessmentNotFoundException has been in edu-services forever. To the grading service, Assignment = Assessment. Go figure.

**Pull request merged**